File: content/snippets/scala2/interview_lcci_01_04_palindrome_permutation.scala
object Lcci0104PalindromePermutation {
  def canPermutePalindrome(s: String): Boolean = {
    val freq = scala.collection.mutable.Map.empty[Char, Int].withDefaultValue(0)

    s.toLowerCase.foreach { ch =>
      if (!ch.isWhitespace) {
        freq.update(ch, freq(ch) + 1)
      }
    }

    freq.values.count(_ % 2 == 1) <= 1
  }

  def main(args: Array[String]): Unit = {
    println(s"tact coa -> ${canPermutePalindrome("tact coa")}")
    println(s"daily -> ${canPermutePalindrome("daily")}")
  }
}

File: content/snippets/python/interview_lcci_01_04_palindrome_permutation.py
def can_permute_palindrome(s: str) -> bool:
    counts: dict[str, int] = {}
    for ch in s.lower():
        if ch.isspace():
            continue
        counts[ch] = counts.get(ch, 0) + 1
    odd = sum(1 for c in counts.values() if c % 2 == 1)
    return odd <= 1


def main() -> None:
    print(f"tact coa -> {can_permute_palindrome('tact coa')}")
    print(f"daily -> {can_permute_palindrome('daily')}")


if __name__ == "__main__":
    main()
